### DISCUSSION TITLE
WPF CefSharp.BrowserSubprocess on window close

### DISCUSSION THREAD

**Discussion by linnuser** [cefsharp/CefSharp#4564] [General]
I have a WPF application that has multiple windows. When I close a window with a ChromiumWebBrowser on some of the CefSharp.BrowserSubprocess remain running. When I close the application all the processes run.

I want to stop the processes because if you keep opening and closing the windows, memory usage is increasing.

In the close event of my windows, I have the following:

```
        MyGrid.Children.Remove(Browser);
        MyBrowser.Dispose();
```

If I add the following all the processes do disappear, however, I can't create a new window with a ChromiumWebBrowser as Initilize can not be called again.

```
        MyGrid.Children.Remove(Browser);
        MyBrowser.Dispose();
        CefSharp.Cef.Shutdown();
```

Any ideas on what I can do?

**Comment by linnuser** [comment]
Ok I have a minimal example:

**MainWindow.xaml**
```
<Window x:Class="WpfApp1.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:WpfApp1"
        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800">
    <Grid>
        <Button Content="Button" HorizontalAlignment="Left" Margin="245,172,0,0" VerticalAlignment="Top" Click="Button_Click"/>

    </Grid>
</Window>
```

**MainWindow.xaml.cs**
```
using System.Windows;

namespace WpfApp1
{
    /// <summary>
    /// Interaction logic for MainWindow.xaml
    /// </summary>
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
        }

        private void Button_Click(object sender, RoutedEventArgs e)
        {
            var p = new HBrowserWindow();

            p.Show();
        }
    }
}
```

**HBwoserWindow.xaml**
```
<Window x:Class="WpfApp1.HBrowserWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:wpf="clr-namespace:CefSharp.Wpf;assembly=CefSharp.Wpf"
        xmlns:local="clr-namespace:WpfApp1"
        mc:Ignorable="d"
        Title="HBrowserWindow" Height="450" Width="800"
        Closing="OnClosing">
    <Grid x:Name="Grid">
        <Grid.ColumnDefinitions>
            <ColumnDefinition/>
        </Grid.ColumnDefinitions>
        <wpf:ChromiumWebBrowser
            x:Name="Browser"
            Address="http://www.google.co.uk" />
    </Grid>
</Window>
```
**HBwoserWindow.xaml.cs**
```
using System.ComponentModel;
using System.Windows;
using System.Windows.Controls;

namespace WpfApp1
{
    /// <summary>
    /// Interaction logic for HBrowserWindow.xaml
    /// </summary>
    public partial class HBrowserWindow : Window
    {
        public HBrowserWindow()
        {
            InitializeComponent();
        }

        private void OnClosing(object sender, CancelEventArgs e)
        {
            Grid.Children.Remove(Browser);
            Browser.Dispose();
        }
    }
}
```

When I click the button I get the following:

![image](https://github.com/cefsharp/CefSharp/assets/43940073/02f4715a-90fa-405a-99aa-f9e36dc2adc2)

When I close the HBwoserWindow I still have this:

![image](https://github.com/cefsharp/CefSharp/assets/43940073/d35d4eba-6bc1-44a3-9a1e-3d3e5d480133)


A CefSharp.BrowserSubprocess process has closed but 4 are left open. I do not want to call shutdown as I may want to press the button again and create a new HBwoserWindow .

**Comment by amaitland** [comment]
> A CefSharp.BrowserSubprocess process has closed but 4 are left open

This is expected, see https://github.com/cefsharp/CefSharp/wiki/General-Usage#processes

`Chromium` controls these processes, long as they are all closed when `Cef.Shutdown` is called then everything is working as I would expect. 

If you deem this behaviour as unacceptable then your options are:
- Look for a different browser control that runs `Chromium` in a separate process.
- Run `CefSharp` in a separate process e.g. https://github.com/cefsharp/CefSharp.OutOfProcess (Only `HwndHost` implementation is supported currently which is equivalent to `CefSharp.Wpf.HwndHost`, there's no equivalent of `CefSharp.Wpf` yet).

---

You can disable the spare render process, and have `Chromium` run the `network` service in process, generally speaking I don't recommend messing with these options.

**Comment by linnuser** [comment]
> Chromium controls these processes, long as they are all closed when Cef.Shutdown is called then everything is working as I would expect.

Cef.Shutdown works great.

> Run CefSharp in a separate process e.g. https://github.com/cefsharp/CefSharp.OutOfProcess (Only HwndHost implementation is supported currently which is equivalent to CefSharp.Wpf.HwndHost, there's no equivalent of CefSharp.Wpf yet).

I will give this a shot. Thank you :)
